Comorphisms/CASL2Prenex.hs

```diff
@@ -0,0 +1,150 @@
+{-# LANGUAGE MultiParamTypeClasses, TypeSynonymInstances, FlexibleInstances #-}
+{- |
+Module      :  $Header$
+Description :  prenex normal form for sentences of a CASL theory
+Copyright   :  (c) Mihai Codescu, 2016
+License     :  GPLv2 or higher, see LICENSE.txt
+
+Maintainer  :  codescu@iws.cs.uni-magdeburg.de
+Stability   :  provisional
+Portability :  non-portable (imports Logic.Comorphism)
+
+-}
+
+module Comorphisms.CASL2Prenex where
+
+import Logic.Logic
+import Logic.Comorphism
+
+import CASL.Logic_CASL
+import CASL.AS_Basic_CASL
+import CASL.Sign
+import CASL.Morphism
+import CASL.Sublogic as SL hiding (bottom)
+import CASL.Induction
+import CASL.Quantification
+
+import Common.Result
+import Common.Id
+import qualified Data.Set as Set
+
+import Common.AS_Annotation
+import Common.ProofTree
+
+data CASL2Prenex = CASL2Prenex deriving Show
+
+instance Language CASL2Prenex where
+    language_name CASL2Prenex = "CASL2Prenex"
+
+instance Comorphism CASL2Prenex
+               CASL CASL_Sublogics
+               CASLBasicSpec CASLFORMULA SYMB_ITEMS SYMB_MAP_ITEMS
+               CASLSign
+               CASLMor
+               Symbol RawSymbol ProofTree
+               CASL CASL_Sublogics
+               CASLBasicSpec CASLFORMULA SYMB_ITEMS SYMB_MAP_ITEMS
+               CASLSign
+               CASLMor
+               Symbol RawSymbol ProofTree where
+    sourceLogic CASL2Prenex = CASL
+    sourceSublogic CASL2Prenex = SL.caslTop --TODO
+    targetLogic CASL2Prenex = CASL
+    mapSublogic CASL2Prenex sl = Just sl -- TODO
+    map_theory CASL2Prenex = mapTheory
+    map_morphism CASL2Prenex = error "nyi"
+    map_sentence CASL2Prenex _ _ = error "nyi"
+    map_symbol CASL2Prenex _ = Set.singleton . id
+    has_model_expansion CASL2Prenex = True -- check
+    is_weakly_amalgamable CASL2Prenex = True --check
+
+-- remove equivalences and implications,
+-- turn arbitrary junctions into binary ones
+preprocessSen :: CASLFORMULA -> CASLFORMULA
```

Comment:
I prefer a prenex normal form computation that leaves the structure of the formula intact as much as possible. Concerning implication and equivalence, see p.10 of http://theo.cs.uni-magdeburg.de/lehre/lehre13w/logik/vl14.pdf
Concerning n-ary conjunctions and disjunctions: they can be handeled analogously to the binary case. So the only needed preprocessing is replacing equivalences by conjunctions of implications.